### PR TITLE
Add pending program status for IPs in CNS.

### PR DESCRIPTION
**Pull request by csfmomo**
*Add pending program status for IPs in CNS.
*Add logic structure of how to update program status.

feat: 
Add logic structure for CNS udpate NC secondary IP to pending program before available

**Reason for Change**:
Pending program is a status to track that IP is available in NC but not in NMAgent side yet. IP is available when NMAgent is also ready. 

**Issue Fixed**:
Prevent CNS allocate pending program IP to pod.

### CODE REVIEW

File: cns/NetworkContainerContract.go [hunk 1]
```diff
@@ -73,6 +74,7 @@ const (
 // CreateNetworkContainerRequest specifies request to create a network container or network isolation boundary.
 type CreateNetworkContainerRequest struct {
 	Version                    string
+	NCVersion                  string
```
[the remark this answers is not in the document]
Comment (neaggarwMS) [reply cns/NetworkContainerContract.go:77]: We should define it as an int

File: cns/requestcontroller/kubecontroller/crdtranslator.go [hunk 2]
```diff
@@ -36,7 +36,7 @@ func CRDStatusToNCRequest(crdStatus nnc.NodeNetworkConfigStatus) (cns.CreateNetw
 		ncRequest.SecondaryIPConfigs = make(map[string]cns.SecondaryIPConfig)
 		ncRequest.NetworkContainerid = nc.ID
 		ncRequest.NetworkContainerType = cns.Docker
-		ncRequest.Version = nc.Version
+		ncRequest.NCVersion = nc.Version
```
Comment (neaggarwMS) [inline cns/requestcontroller/kubecontroller/crdtranslator.go:39]: Why is this changed? It should be set to ncRequest.Version only

File: cns/ipampoolmonitor/ipampoolmonitor.go [hunk 3]
```diff
@@ -64,12 +64,14 @@ func (pm *CNSIPAMPoolMonitor) Start(ctx context.Context, poolMonitorRefreshMilli
...
 func (pm *CNSIPAMPoolMonitor) Reconcile() error {
 	cnsPodIPConfigCount := len(pm.cns.GetPodIPConfigState())
+	pendingProgramCount := len(pm.cns.GetPendingProgramIPConfigs()) // TODO: add pending program count to real cns
 	allocatedPodIPCount := len(pm.cns.GetAllocatedIPConfigs())
 	pendingReleaseIPCount := len(pm.cns.GetPendingReleaseIPConfigs())
 	availableIPConfigCount := len(pm.cns.GetAvailableIPConfigs()) // TODO: add pending allocation count to real cns
 	freeIPConfigCount := pm.cachedNNC.Spec.RequestedIPCount - int64(allocatedPodIPCount)
 
-	logger.Printf("[ipam-pool-monitor] Checking pool for resize, Pool Size: %v, Goal Size: %v, BatchSize: %v, MinFree: %v, MaxFree:%v, Allocated: %v, Available: %v, Pending Release: %v, Free: %v", cnsPodIPConfigCount, pm.cachedNNC.Spec.RequestedIPCount, pm.scalarUnits.BatchSize, pm.MinimumFreeIps, pm.MaximumFreeIps, allocatedPodIPCount, availableIPConfigCount, pendingReleaseIPCount, freeIPConfigCount)
+	logger.Printf("[ipam-pool-monitor] Checking pool for resize, Pool Size: %v, Goal Size: %v, BatchSize: %v, MinFree: %v, MaxFree:%v, Allocated: %v, Available: %v, Pending Release: %v, Free: %v, Pending Program: %v",
```
Comment (neaggarwMS) [inline cns/ipampoolmonitor/ipampoolmonitor.go:73]: We were getting too much noise from this log, @matmerr we should check how to reduce the noise.

File: cns/restserver/util.go [hunk 4]
```diff
@@ -236,37 +237,60 @@ func (service *HTTPRestService) updateIpConfigsStateUntransacted(req cns.CreateN
...
-	// add ipconfigs to state
-	for ipId, ipconfig := range ipconfigs {
-		// if this IPConfig already exists in the map, then ignore as this is an idempotent state
-		if _, exists := service.PodIPConfigState[ipId]; exists {
-			continue
-		}
+func (service *HTTPRestService) addIPConfigStateUntransacted(ncVersion, ncId string, ipconfigs map[string]cns.SecondaryIPConfig) {
+
+	existingNCStatus, _ := service.state.ContainerStatus[ncId]
+	existingNCVersion, _ := strconv.Atoi(existingNCStatus.CreateNetworkContainerRequest.NCVersion)
```
Comment (neaggarwMS) [inline cns/restserver/util.go:251]: This wont return the expected value,  NCRequest is already overwritten in the caller API (saveNetworkContainerGoalState, line 120). You need to pass existing Version (not NCVersion) from the called like we are passing existingIpconfigs

File: cns/restserver/util.go [hunk 5]
```diff
@@ -236,37 +237,60 @@ func (service *HTTPRestService) updateIpConfigsStateUntransacted(req cns.CreateN
...
+	nmAgentNCVersion := getNCVersionFromNMAgent(ncId)
```
Comment (neaggarwMS) [inline cns/restserver/util.go:252]: I actually recommend to move it to background or make a call in a go routing and wait for it.

File: cns/restserver/util.go [hunk 6]
```diff
@@ -236,37 +237,60 @@ func (service *HTTPRestService) updateIpConfigsStateUntransacted(req cns.CreateN
...
-func (service *HTTPRestService) addIPConfigStateUntransacted(ncId string, ipconfigs map[string]cns.SecondaryIPConfig) {
-	// add ipconfigs to state
-	for ipId, ipconfig := range ipconfigs {
-		// if this IPConfig already exists in the map, then ignore as this is an idempotent state
-		if _, exists := service.PodIPConfigState[ipId]; exists {
-			continue
-		}
+func (service *HTTPRestService) addIPConfigStateUntransacted(ncVersion, ncId string, ipconfigs map[string]cns.SecondaryIPConfig) {
+
+	existingNCStatus, _ := service.state.ContainerStatus[ncId]
```
Comment (neaggarwMS) [inline cns/restserver/util.go:250]: Again this is already updated in the caller

File: cns/restserver/util.go [hunk 7]
```diff
@@ -236,37 +237,60 @@ func (service *HTTPRestService) updateIpConfigsStateUntransacted(req cns.CreateN
...
+				OrchestratorContext: nil,
+			}
+
+			service.PodIPConfigState[ipId] = ipconfigStatus
 
-		service.PodIPConfigState[ipId] = ipconfigStatus
+			// Todo Update batch API and maintain the count
 
-		// Todo Update batch API and maintain the count
+			existingNCStatus.CreateNetworkContainerRequest.NCVersion = strconv.Itoa(newNCVersion)
```
Comment (neaggarwMS) [inline cns/restserver/util.go:275]: you dont need this as it is already updated above

File: cns/restserver/util.go [hunk 8]
```diff
@@ -236,37 +237,60 @@ func (service *HTTPRestService) updateIpConfigsStateUntransacted(req cns.CreateN
...
+			// Unconmment the following command when getNCVersionFromNMAgent function finished.
+			// existingNCStatus.CreateNetworkContainerRequest.NCVersion = strconv.Itoa(nmAgentNCVersion)
+		} else {
+			// todo, update IPs in ipconfigs which is not in the map to pending program.
+		}
 	}
 }
 
+// To do, complete this logic according to getNetworkContainerStatus
+func getNCVersionFromNMAgent(ncid string) int {
```
Comment (neaggarwMS) [inline cns/restserver/util.go:289]: Please add the logic to get NMAgent ver in ImdsClientInterface

File: cns/restserver/util.go [hunk 9]
```diff
@@ -236,37 +237,60 @@ func (service *HTTPRestService) updateIpConfigsStateUntransacted(req cns.CreateN
...
 // If the IP is already added then it will be an idempotent call. Also note, caller will
 // acquire/release the service lock.
-func (service *HTTPRestService) addIPConfigStateUntransacted(ncId string, ipconfigs map[string]cns.SecondaryIPConfig) {
-	// add ipconfigs to state
-	for ipId, ipconfig := range ipconfigs {
-		// if this IPConfig already exists in the map, then ignore as this is an idempotent state
-		if _, exists := service.PodIPConfigState[ipId]; exists {
-			continue
-		}
+func (service *HTTPRestService) addIPConfigStateUntransacted(ncVersion, ncId string, ipconfigs map[string]cns.SecondaryIPConfig) {
```
Comment (neaggarwMS) [inline cns/restserver/util.go:248]: I think in this function you also need to pass existingIpconfigs so that you can keep those IPs as available and mark the new ones in PendingProgramming State

File: cns/NetworkContainerContract.go [hunk 10]
```diff
@@ -61,6 +61,7 @@ const (
 	Available      = "Available"
 	Allocated      = "Allocated"
 	PendingRelease = "PendingRelease"
+	PendingProgram = "PendingProgram"
```
Comment (neaggarwMS) [inline cns/NetworkContainerContract.go:64]: can we rename it to PendingProgramming

File: cns/NetworkContainerContract.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (neaggarwMS) [reply cns/NetworkContainerContract.go:77]: Note: we need to persist the ProgrammedVersion so that we can avoid calling NMAgent again if CNS crashes

File: cns/restserver/util.go [hunk 11]
```diff
@@ -236,37 +237,60 @@ func (service *HTTPRestService) updateIpConfigsStateUntransacted(req cns.CreateN
 	}
 
 	// Add the newIpConfigs, ignore if ip state is already in the map
-	service.addIPConfigStateUntransacted(req.NetworkContainerid, newIPConfigs)
+	service.addIPConfigStateUntransacted(req.NCVersion, req.NetworkContainerid, newIPConfigs)
```
Comment (neaggarwMS) [inline cns/restserver/util.go:240]: Can you add some logs to updateIpConfigsStateUntransacted function, it is missing logs overall. Specially when we are updating the state or if this api is returning error

**Review by neaggarwMS — CHANGES_REQUESTED** [review]

File: cns/NetworkContainerContract.go (hunk 10 above)
Comment (csfmomo) [reply cns/NetworkContainerContract.go:64]: Sure.

File: cns/requestcontroller/kubecontroller/crdtranslator.go (hunk 2 above)
Comment (csfmomo) [reply cns/requestcontroller/kubecontroller/crdtranslator.go:39]: Will change it back to ncRequest.Version. 
I didn't reuse the Version since the usage of Version is not clear in CNS previously. There are Version, which compared with VMVersion and HostVersion which record NMA version. After double checking code, resusing Version should be fine.

File: cns/restserver/util.go (hunk 11 above)
Comment (csfmomo) [reply cns/restserver/util.go:240]: Sure.

File: cns/restserver/util.go (hunk 9 above)
Comment (csfmomo) [reply cns/restserver/util.go:248]: Yes, existingSecondaryIPConfigs is needed.

File: cns/restserver/util.go (hunk 6 above)
Comment (csfmomo) [reply cns/restserver/util.go:250]: Will update.

File: cns/restserver/util.go (hunk 4 above)
Comment (csfmomo) [reply cns/restserver/util.go:251]: Will update.

File: cns/restserver/util.go (hunk 5 above)
Comment (csfmomo) [reply cns/restserver/util.go:252]: No matter whether we query NMAgent in background or not, we should compare NMAgent's version every time when NC is updated. I don't think NC update will be too frequent to bug NMAgent.

File: cns/restserver/util.go (hunk 8 above)
Comment (csfmomo) [reply cns/restserver/util.go:289]: getNetworkContainerStatus is using imdsClient.GetNetworkContainerInfoFromHost. I think GetNetworkContainerInfoFromHost covers most of this part logic. I can double check. Will update there.

File: cns/restserver/util.go (hunk 7 above)
Comment (csfmomo) [reply cns/restserver/util.go:275]: OK

File: cns/NetworkContainerContract.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (csfmomo) [reply cns/NetworkContainerContract.go:77]: 1. I'll remove NCVersion. No need to record ProgrammedVersion in CreateNetworkContainerRequest since similar info got record in containerstatus and containerstatus include CreateNetworkContainerRequest. I added previously since the usage of Version is not clear in CNS. There are VMVersion and hostVersion which compared with this Version. So I added another explicite NCVersion to ensure AKS-Swift usage. However, after double checking every usage of Version, I can confirm Version meant NCVersion. I'll reuse it.
2. I'll remove NCVersion. If we want to change Version to int, we can open a separate PR to change all the existing Version.
3. ProgrammedVersion is persisted in containerstatus already.

File: cns/ipampoolmonitor/ipampoolmonitor.go (hunk 3 above)
Comment (csfmomo) [reply cns/ipampoolmonitor/ipampoolmonitor.go:73]: How about I remain this log in this PR but create a separate PR to reduce log noise.

File: cns/restserver/util.go [hunk 12]
```diff
@@ -235,29 +236,52 @@ func (service *HTTPRestService) updateIpConfigsStateUntransacted(req cns.CreateN
...
+	existingNCVersion, _ = strconv.Atoi(existingNCVersionInString)
+	// Wait for Ashvin's thread update which querying NMAgent to get latest NC version and save it locally.
+	nmAgentNCVersion := getNCVersionFromNMAgent(req.NetworkContainerid)
+	if nmAgentNCVersion >= newNCVersion {
+		service.addIPConfigStateUntransacted(cns.Available, cns.Available, req.NetworkContainerid, newIPConfigs)
+	} else {
+		if nmAgentNCVersion >= existingNCVersion {
+			service.addIPConfigStateUntransacted(cns.Available, cns.PendingProgramming, req.NetworkContainerid, newIPConfigs)
+		} else {
+			service.addIPConfigStateUntransacted(cns.PendingProgramming, cns.PendingProgramming, req.NetworkContainerid, newIPConfigs)
```
Comment (neaggarwMS) [inline cns/restserver/util.go:255]: Why we need to move existing IPs back to PendingState if they are already available?

File: cns/restserver/util.go [hunk 13]
```diff
@@ -235,29 +236,52 @@ func (service *HTTPRestService) updateIpConfigsStateUntransacted(req cns.CreateN
...
+	// Add the newIpConfigs for 3 different conditions
+	// 1. If NC version from NMAgent is larger or same as new NC version, add new IP as available.
+	// 2. If NC version from NMAgent is larger or same as existing NC version, set existing pending programming IP as available and add new IP as pending programming.
+	// 3. If NC version from NMAgent is smaller than existing NC version, add new IP as pending programming.
+	var newNCVersion int
+	var existingNCVersion int
+	newNCVersion, _ = strconv.Atoi(req.Version)
+	existingNCVersion, _ = strconv.Atoi(existingNCVersionInString)
+	// Wait for Ashvin's thread update which querying NMAgent to get latest NC version and save it locally.
+	nmAgentNCVersion := getNCVersionFromNMAgent(req.NetworkContainerid)
```
Comment (neaggarwMS) [inline cns/restserver/util.go:248]: Again, I dont think we should be able to call NMAgent while acquiring the Service lock. This function should compare the NCVersion with existing NMAgent version (persisted in ContainerStatus) and if it is higher than set the ips to PendingProgramming and exit.

File: cns/restserver/util.go [hunk 14]
```diff
@@ -112,9 +113,17 @@ func (service *HTTPRestService) saveNetworkContainerGoalState(req cns.CreateNetw
 	existingNCStatus, ok := service.state.ContainerStatus[req.NetworkContainerid]
 	var hostVersion string
 	var existingSecondaryIPConfigs map[string]cns.SecondaryIPConfig //uuid is key
+	var existingNCVersion string
 	if ok {
 		hostVersion = existingNCStatus.HostVersion
```
Comment (neaggarwMS) [inline cns/restserver/util.go:118]: What is HostVersion? Are you sure that is NCVersion in NMAgent?

File: cns/restserver/util.go [hunk 15]
```diff
@@ -112,9 +113,17 @@ func (service *HTTPRestService) saveNetworkContainerGoalState(req cns.CreateNetw
...
 		existingSecondaryIPConfigs = existingNCStatus.CreateNetworkContainerRequest.SecondaryIPConfigs
+		existingNCVersion = existingNCStatus.VMVersion
```
Comment (neaggarwMS) [inline cns/restserver/util.go:120]: Can we pick it from existingNCStatus.CreateNetworkContainerRequest.Version? Dont know if we set the VMVersion for sure in AKS scenario

File: cns/restserver/util.go [hunk 16]
```diff
@@ -112,9 +113,17 @@ func (service *HTTPRestService) saveNetworkContainerGoalState(req cns.CreateNetw
...
+	} else {
+		// Host version is the NC version from NMAgent, set it -2 to indicate no result from NMAgent yet.
+		// Existing NC status will be set as -1 to indicate no version exist yet.
+		// Intentionally set default host version smaller than existing NC version.
+		hostVersion = "-2"
```
Comment (neaggarwMS) [inline cns/restserver/util.go:125]: Are we sure hostVersion represents NC version from NMagent?

File: cns/restserver/util.go (hunk 14 above)
Comment (neaggarwMS) [reply cns/restserver/util.go:118]: Looks like it is

File: cns/restserver/util.go (hunk 16 above)
Comment (neaggarwMS) [reply cns/restserver/util.go:125]: Why -2 or -1. We must never receive NC GS with Version 0. Set hostVersion to 0

File: cns/restserver/util.go [hunk 17]
```diff
@@ -112,9 +113,17 @@ func (service *HTTPRestService) saveNetworkContainerGoalState(req cns.CreateNetw
...
+		existingNCVersion = "-1"
```
Comment (neaggarwMS) [inline cns/restserver/util.go:126]: Again, set this to 0, in case it is not set. We must confirm the version we are getting from CRD (must always be non zero)

File: cns/restserver/util.go (hunk 13 above)
Comment (neaggarwMS) [reply cns/restserver/util.go:248]: Instead of calling getNCVersionFromNMAgent, you should pull HostVersion from ContainerStatus.HostVersion. That is only required in case updated NCVersion is same as HostVersion, in that case we will set the ips to Available state, else Ips are in PendingState if not already set

File: cns/restserver/util.go [hunk 18]
```diff
@@ -235,29 +236,52 @@ func (service *HTTPRestService) updateIpConfigsStateUntransacted(req cns.CreateN
...
+	}
 
 	return 0, ""
 }
 
 // addIPConfigStateUntransacted adds the IPConfigs to the PodIpConfigState map with Available state
 // If the IP is already added then it will be an idempotent call. Also note, caller will
 // acquire/release the service lock.
-func (service *HTTPRestService) addIPConfigStateUntransacted(ncId string, ipconfigs map[string]cns.SecondaryIPConfig) {
+func (service *HTTPRestService) addIPConfigStateUntransacted(updatedExistingIPCNSStatus, newIPCNSStatus, ncId string, ipconfigs map[string]cns.SecondaryIPConfig) {
```
Comment (neaggarwMS) [inline cns/restserver/util.go:270]: No need to pass updatedExistingIPCNSStatus. In case existing IP is in state Pending, set it to newIPCNSStatus agnostically

Comment (neaggarwMS) [reply cns/restserver/util.go:270]: can we rename this function to addOrUpdateIPConfigStateUntransacted?

**Review by neaggarwMS — CHANGES_REQUESTED** [review]

File: cns/restserver/util.go (hunk 12 above)
Comment (csfmomo) [reply cns/restserver/util.go:255]: 1. We are NOT make an outboud call. getNCVersionFromNMAgent didn't have implementation yet. It could be the function which fetch NMAgent version locally from what background thread got and saved.
2. We not not moving existing IPs back to Pending State.

File: cns/restserver/util.go (hunk 13 above)
Comment (csfmomo) [reply cns/restserver/util.go:248]: getNCVersionFromNMAgent didn't have any implementation yet. It could be the function which fetch NMAgent version locally from what background thread got and saved.
Whether it's get from ContainerStatus.HostVersion or not, it depends on Ashvin's PR  that where will he save the NMAgent version.

File: cns/restserver/util.go (hunk 14 above)
Comment (csfmomo) [reply cns/restserver/util.go:118]: It it.

File: cns/restserver/util.go [hunk 19]
```diff
@@ -235,29 +236,52 @@ func (service *HTTPRestService) updateIpConfigsStateUntransacted(req cns.CreateN
...
 	// add ipconfigs to state
 	for ipId, ipconfig := range ipconfigs {
-		// if this IPConfig already exists in the map, then ignore as this is an idempotent state
-		if _, exists := service.PodIPConfigState[ipId]; exists {
+		if existingPodIpConfigStatus, exists := service.PodIPConfigState[ipId]; exists {
+			if existingPodIpConfigStatus.State == cns.PendingProgramming {
+				existingPodIpConfigStatus.State = updatedExistingIPCNSStatus
```
Comment (neaggarwMS) [inline cns/restserver/util.go:275]: Can you add a log in case existing state is getting updated?

File: cns/restserver/util.go (hunk 16 above)
Comment (csfmomo) [reply cns/restserver/util.go:125]: 1. I'm sure. I checked every place where host version got implemented, set value and referenced. 
2. Can't set it 0. NC version started from 0 in DNC code.

File: cns/restserver/util.go (hunk 17 above)
Comment (csfmomo) [reply cns/restserver/util.go:126]: 2 reasons can't set it to 0.
1.  NC version started from 0 in DNC code.
2. Host version should not be the same as existing NC version in default mode.

File: cns/restserver/util.go (hunk 18 above)
Comment (csfmomo) [reply cns/restserver/util.go:270]: 1. The reason why I have updatedExistingIPCNSStatus is when updated NMAgent version is the same as NC version, updatedExistingIPCNSStatus will be available while newIPCNSStatus is pending programming. Check the comment in 241 line in util.go
2. Yes, I can rename.

File: cns/restserver/util.go (hunk 19 above)
Comment (csfmomo) [reply cns/restserver/util.go:275]: I can.

File: cns/restserver/util.go (hunk 15 above)
Comment (neaggarwMS) [reply cns/restserver/util.go:120]: We dont need existingNCVersion

Comment (csfmomo) [reply cns/restserver/util.go:120]: We need existing NC version if we want to update existing IPs from pending programming to available. Only when host version is the same as exsting NC version, CNS can update existing IPs from pending programming to available.
If we want to move pending -> available logic to background thread, then no need of existing NC version.

File: cns/restserver/util.go [hunk 20]
```diff
@@ -235,31 +230,29 @@ func (service *HTTPRestService) updateIpConfigsStateUntransacted(req cns.CreateN
 		}
 	}
 
-	// Add the newIpConfigs, ignore if ip state is already in the map
-	service.addIPConfigStateUntransacted(req.NetworkContainerid, newIPConfigs)
+	service.addIPConfigStateUntransacted(cns.Available, req.NetworkContainerid, newIPConfigs)
```
Comment (neaggarwMS) [inline cns/restserver/util.go:233]: When are you setting them as pending?

**Review by neaggarwMS — CHANGES_REQUESTED** [review]
Left one minor comment.

File: cns/fakes/imdsclientfake.go [hunk 21]
```diff
@@ -47,3 +47,11 @@ func (imdsClient *ImdsClientTest) GetPrimaryInterfaceInfoFromMemory() (*imdsclie
 
 	return imdsClient.GetPrimaryInterfaceInfoFromHost()
 }
+
+// GetNMagentVersion - Mock implementation to return host NMAgent NC version
+// Set it as 0 which is the same as default initial NC version for testing purpose
+func (imdsClient *ImdsClientTest) GetNMagentVersion() int {
+	logger.Printf("[Azure CNS] GetNMagentVersionFromNMAgent")
```
Comment (neaggarwMS) [inline cns/fakes/imdsclientfake.go:54]: Update the log to get the NC version from NMAgent

File: cns/imdsclient/api.go [hunk 22]
```diff
@@ -74,4 +74,5 @@ type ImdsClientInterface interface {
 	GetNetworkContainerInfoFromHost(networkContainerID string, primaryAddress string, authToken string, apiVersion string) (*ContainerVersion, error)
 	GetPrimaryInterfaceInfoFromHost() (*InterfaceInfo, error)
 	GetPrimaryInterfaceInfoFromMemory() (*InterfaceInfo, error)
+	GetNMagentVersion() int
```
Comment (neaggarwMS) [inline cns/imdsclient/api.go:77]: Rename the function name GetNetworkContainerInfoFromHost and pass all the parameters networkContainerID string, primaryAddress string, apiVersion string)

**Review by neaggarwMS — CHANGES_REQUESTED** [review]

**Review by neaggarwMS — APPROVED** [review]

**Review by neaggarwMS — APPROVED** [review]
